A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWork

<!--- 
Please use our bug report template to report problems with something that has never worked.  
Regressions reports are greatly appreciated during our RC phase and before a final release.
-->

<!--- Provide a brief summary of the issue in the title above -->
When using Gradle 5.6 to build our Kotlin project we experience build failures with the message given in the title.  
This happens reproducibly for our complete build (`./gradlew clean build --no-build-cache --rerun-tasks`).
The failing tasks have no problem when we execute them alone (`./gradlew clean :our_module:compileKotlin`) .
With Gradle 5.5.1 we don't have this problem.

I'm posting this problem here, because it is Gradle code that throws the initial `IllegalArgumentException`.

### Expected Behavior
<!--- Tell us what should happen -->
Gradle should compile the Kotlin sources successfully like in version 5.5.1 

### Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
The `compileKotlin` task fails with the following exception:

```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':common:compileTestKotlin'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:166)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.accept(ExecuteActionsTaskExecuter.java:163)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:191)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:156)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:108)
        at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionOutputsTaskExecuter.execute(ResolveBeforeExecutionOutputsTaskExecuter.java:67)
        at org.gradle.api.internal.tasks.execution.ResolveAfterPreviousExecutionStateTaskExecuter.execute(ResolveAfterPreviousExecutionStateTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:94)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:95)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.workers.internal.DefaultWorkerExecutor$WorkExecutionException: A failure occurred while executing org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWork
        at org.gradle.workers.internal.DefaultWorkerExecutor$WorkItemExecution.waitForCompletion(DefaultWorkerExecutor.java:372)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:142)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForItemsAndGatherFailures(DefaultAsyncWorkTracker.java:94)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForAll(DefaultAsyncWorkTracker.java:80)
        at org.gradle.internal.work.DefaultAsyncWorkTracker.waitForCompletion(DefaultAsyncWorkTracker.java:68)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$5.run(ExecuteActionsTaskExecuter.java:484)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:461)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:444)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:93)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:237)
        at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$0(ExecuteStep.java:32)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:32)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:58)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:35)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:48)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:33)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:39)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
        at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:35)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:45)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:31)
        at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:208)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:70)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:45)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:43)
        at org.gradle.internal.execution.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:32)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:96)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:89)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:54)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:38)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:90)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:48)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:69)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:47)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:140)
        ... 34 more
Caused by: org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWork.
        at org.gradle.internal.instantiation.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:54)
        at org.gradle.workers.internal.AdapterWorkAction.execute(AdapterWorkAction.java:48)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:47)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1$1.create(NoIsolationWorkerFactory.java:65)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1$1.create(NoIsolationWorkerFactory.java:61)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:98)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.execute(NoIsolationWorkerFactory.java:61)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:44)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:41)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.workers.internal.AbstractWorker.executeWrappedInBuildOperation(AbstractWorker.java:41)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1.execute(NoIsolationWorkerFactory.java:56)
        at org.gradle.workers.internal.DefaultWorkerExecutor$3.call(DefaultWorkerExecutor.java:212)
        at org.gradle.workers.internal.DefaultWorkerExecutor$3.call(DefaultWorkerExecutor.java:206)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runExecution(DefaultConditionalExecutionQueue.java:215)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runBatch(DefaultConditionalExecutionQueue.java:164)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.run(DefaultConditionalExecutionQueue.java:131)
        ... 3 more
Caused by: java.lang.IllegalArgumentException: Unable to determine constructor argument #1: value org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWorkArguments@26eb6d07 not assignable to class org.jetbrains.kotlin.compilerRunner.GradleKotlinCompilerWorkArguments
        at org.gradle.internal.instantiation.DependencyInjectingInstantiator.verifyParameters(DependencyInjectingInstantiator.java:130)
        at org.gradle.internal.instantiation.DependencyInjectingInstantiator.convertParameters(DependencyInjectingInstantiator.java:104)
        at org.gradle.internal.instantiation.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:47)
        ... 25 more
```

### Context
<!--- Which version did this stop working with? Which version do you know it works? -->
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
Worked with Gradle 5.5.1, stopped working with Gradle 5.6.
[edit] We're hitting this issue with both AdoptOpenJDK (build 11.0.4+11) and Oracle 12.0.2. 

### Steps to Reproduce 
<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -->
As it would be quite laborious to create a self-contained example and I'm not sure whether it's a Gradle or a Kotlin Gradle plugin issue, I don't have one (yet). 

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL:
https://scans.gradle.com/s/c43rxmm3mqbhm

--- a/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
+++ b/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
@@ -1,0 +1,271 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.workers.internal
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.workers.fixtures.WorkerExecutorFixture
+import spock.lang.Unroll
+
+class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpec {
+    WorkerExecutorFixture fixture = new WorkerExecutorFixture(temporaryFolder)
+    def plugin = testDirectory.createDir("plugin")
+    def build1 = testDirectory.createDir("build1")
+
+    @Unroll
+    def "can use worker api with composite builds using #pluginId"() {
+        settingsFile << """
+            includeBuild "plugin"
+            includeBuild "build1"
+        """
+
+        withFileHelperInPluginBuild()
+        withLegacyRunnableAndTaskInPluginBuild()
+        withTypedWorkerPluginInPluginBuild()
+
+        build1.file("build.gradle") << """
+            buildscript {
+                repositories {
+                    ${jcenterRepository()}
+                }
+
+                dependencies {
+                    classpath "org.apache.commons:commons-math:2.2"
+                }
+            }
+
+            plugins {
+                id "java"
+                id "org.gradle.test.${pluginId}"
+            }
+
+            group = "org.gradle.test"
+            version = "1.0"
+
+            jar {
+                from runWork
+            }
+        """
+
+        buildFile << """
+            plugins {
+                id "java"
+                id "org.gradle.test.${pluginId}"
+            }
+       
+            dependencies {
+                compile "org.gradle.test:build1:1.0"
+            }
+
+            runWork.dependsOn compileJava
+        """
+
+        expect:
+        succeeds("runWork")
+
+        where:
+        pluginId << [ 'legacy-worker-plugin', 'typed-worker-plugin' ]
+    }
+
+    private void withLegacyRunnableAndTaskInPluginBuild() {
+        plugin.file("src/main/java/LegacyParameter.java") << """
+            import java.io.File;
+            import java.io.Serializable;
+            
+            public class LegacyParameter implements Serializable {
+                private final File outputFile;
+                
+                public LegacyParameter(File outputFile) {
+                    this.outputFile = outputFile;
+                }
+                
+                public File getOutputFile() {
+                    return this.outputFile;
+               }
+            }
+        """
+
+        plugin.file('src/main/java/LegacyRunnable.java') << """
+            import javax.inject.Inject;
+            import java.io.File;
+            import org.gradle.test.FileHelper;
+            
+            public class LegacyRunnable implements Runnable {
+                private File outputFile;
+                
+                @Inject
+                public LegacyRunnable(LegacyParameter parameter) {
+                    this.outputFile = parameter.getOutputFile();
+                }
+                
+                public void run() {
+                    FileHelper.write("foo", outputFile);
+                }
+            }
+        """
+
+        plugin.file('src/main/java/LegacyWorkerTask.java') << """
+            import org.gradle.api.DefaultTask;
+            import org.gradle.api.Action;
+            import org.gradle.api.tasks.*;
+            import org.gradle.workers.*;
+            import javax.inject.Inject;
+            import java.io.File;
+            
+            public class LegacyWorkerTask extends DefaultTask {
+                private final WorkerExecutor workerExecutor;
+                private File outputFile;
+                
+                @Inject
+                public LegacyWorkerTask(WorkerExecutor workerExecutor) {
+                    this.workerExecutor = workerExecutor;
+                } 
+                
+                @TaskAction
+                public void runWork() {
+                    workerExecutor.submit(LegacyRunnable.class, new Action<WorkerConfiguration>() {
+                        public void execute(WorkerConfiguration config) {
+                            config.setIsolationMode(IsolationMode.NONE);
+                            config.params(new LegacyParameter(outputFile));
+                        }
+                    });
+                }
+                
+                @OutputFile
+                File getOutputFile() {
+                    return outputFile;
+                }
+                
+                void setOutputFile(File outputFile) {
+                    this.outputFile = outputFile;
+                }
+            }
+        """
+
+        plugin.file("src/main/java/LegacyWorkerPlugin.java") << """
+            import org.gradle.api.Plugin;
+            import org.gradle.api.Project;
+            import java.io.File;
+            
+            public class LegacyWorkerPlugin implements Plugin<Project> {
+                public void apply(Project project) {
+                    project.getTasks().create("runWork", LegacyWorkerTask.class)
+                        .setOutputFile(new File(project.getBuildDir(), "workOutput"));
+                }
+            }
+        """
+
+        plugin.file("build.gradle") << """
+            apply plugin: "java-gradle-plugin"
+            
+            gradlePlugin {
+                plugins {
+                    LegacyWorkerPlugin {
+                        id = "org.gradle.test.legacy-worker-plugin"
+                        implementationClass = "LegacyWorkerPlugin"
+                    }
+                }
+            }
+        """
+    }
+
+    private void withTypedWorkerPluginInPluginBuild() {
+        plugin.file("src/main/java/TypedParameter.java") << """
+            import org.gradle.workers.WorkParameters;
+            import org.gradle.api.file.RegularFileProperty;
+            
+            public interface TypedParameter extends WorkParameters {
+                RegularFileProperty getOutputFile();
+            }
+        """
+
+        plugin.file("src/main/java/TypedWorkAction.java") << """
+            import org.gradle.workers.WorkAction;
+            import org.gradle.test.FileHelper;
+            
+            abstract public class TypedWorkAction implements WorkAction<TypedParameter> {
+                public void execute() {
+                    FileHelper.write("foo", getParameters().getOutputFile().getAsFile().get());
+                }
+            }
+        """
+
+        plugin.file("src/main/java/TypedWorkerTask.java") << """
+            import org.gradle.api.DefaultTask;
+            import org.gradle.api.Action;
+            import org.gradle.api.file.RegularFileProperty;
+            import org.gradle.api.tasks.*;
+            import org.gradle.workers.*;
+            import javax.inject.Inject;
+            import java.io.File;
+            
+            public class TypedWorkerTask extends DefaultTask {
+                private final WorkerExecutor workerExecutor;
+                private RegularFileProperty outputFile;
+                
+                @Inject
+                public TypedWorkerTask(WorkerExecutor workerExecutor) {
+                    this.workerExecutor = workerExecutor;
+                    this.outputFile = getProject().getObjects().fileProperty();
+                } 
+                
+                @TaskAction
+                public void runWork() {
+                    workerExecutor.noIsolation().submit(TypedWorkAction.class, new Action<TypedParameter>() {
+                        public void execute(TypedParameter parameters) {
+                            parameters.getOutputFile().set(outputFile);
+                        }
+                    });
+                }
+                
+                @OutputFile
+                RegularFileProperty getOutputFile() {
+                    return outputFile;
+                }
+            }
+        """
+
+        plugin.file("src/main/java/TypedWorkerPlugin.java") << """
+            import org.gradle.api.Plugin;
+            import org.gradle.api.Project;
+            import java.io.File;
+            
+            public class TypedWorkerPlugin implements Plugin<Project> {
+                public void apply(Project project) {
+                    project.getTasks().create("runWork", TypedWorkerTask.class)
+                        .getOutputFile().set(new File(project.getBuildDir(), "workOutput"));
+                }
+            }
+        """
+
+        plugin.file("build.gradle") << """
+            apply plugin: "java-gradle-plugin"
+            
+            gradlePlugin {
+                plugins {
+                    TypedWorkerPlugin {
+                        id = "org.gradle.test.typed-worker-plugin"
+                        implementationClass = "TypedWorkerPlugin"
+                    }
+                }
+            }
+        """
+    }
+
+    private void withFileHelperInPluginBuild() {
+        plugin.file("src/main/java/org/gradle/test/FileHelper.java") << fixture.fileHelperClass
+    }
+}

--- a/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
+++ b/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
@@ -42,12 +42,8 @@ class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpe
 
         lib.file("build.gradle") << """
             buildscript {
-                repositories {
-                    ${jcenterRepository()}
-                }
-
                 dependencies {
-                    classpath "org.apache.commons:commons-math:2.2"
+                    classpath files('foo.jar')
                 }
             }
 

--- a/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
+++ b/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
@@ -24,21 +24,23 @@ import spock.lang.Unroll
 class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpec {
     WorkerExecutorFixture fixture = new WorkerExecutorFixture(temporaryFolder)
     def plugin = testDirectory.createDir("plugin")
-    def build1 = testDirectory.createDir("build1")
+    def lib = testDirectory.createDir("lib")
 
     @Unroll
     @Issue("https://github.com/gradle/gradle/issues/10317")
     def "can use worker api with composite builds using #pluginId"() {
         settingsFile << """
+            rootProject.name = "app"
+            
             includeBuild "plugin"
-            includeBuild "build1"
+            includeBuild "lib"
         """
 
         withFileHelperInPluginBuild()
         withLegacyWorkerPluginInPluginBuild()
         withTypedWorkerPluginInPluginBuild()
 
-        build1.file("build.gradle") << """
+        lib.file("build.gradle") << """
             buildscript {
                 repositories {
                     ${jcenterRepository()}
@@ -69,7 +71,7 @@ class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpe
             }
        
             dependencies {
-                compile "org.gradle.test:build1:1.0"
+                compile "org.gradle.test:lib:1.0"
             }
 
             runWork.dependsOn compileJava
@@ -80,7 +82,7 @@ class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpe
 
         and:
         file("build/workOutput").text == "foo"
-        build1.file("build/workOutput").text == "foo"
+        lib.file("build/workOutput").text == "foo"
 
         where:
         pluginId << [ 'legacy-worker-plugin', 'typed-worker-plugin' ]

--- a/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
+++ b/subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorCompositeBuildIntegrationTest.groovy
@@ -18,6 +18,7 @@ package org.gradle.workers.internal
 
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.workers.fixtures.WorkerExecutorFixture
+import spock.lang.Issue
 import spock.lang.Unroll
 
 class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpec {
@@ -26,6 +27,7 @@ class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpe
     def build1 = testDirectory.createDir("build1")
 
     @Unroll
+    @Issue("https://github.com/gradle/gradle/issues/10317")
     def "can use worker api with composite builds using #pluginId"() {
         settingsFile << """
             includeBuild "plugin"
@@ -33,7 +35,7 @@ class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpe
         """
 
         withFileHelperInPluginBuild()
-        withLegacyRunnableAndTaskInPluginBuild()
+        withLegacyWorkerPluginInPluginBuild()
         withTypedWorkerPluginInPluginBuild()
 
         build1.file("build.gradle") << """
@@ -76,11 +78,15 @@ class WorkerExecutorCompositeBuildIntegrationTest extends AbstractIntegrationSpe
         expect:
         succeeds("runWork")
 
+        and:
+        file("build/workOutput").text == "foo"
+        build1.file("build/workOutput").text == "foo"
+
         where:
         pluginId << [ 'legacy-worker-plugin', 'typed-worker-plugin' ]
     }
 
-    private void withLegacyRunnableAndTaskInPluginBuild() {
+    private void withLegacyWorkerPluginInPluginBuild() {
         plugin.file("src/main/java/LegacyParameter.java") << """
             import java.io.File;
             import java.io.Serializable;

--- a/subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java
+++ b/subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java
@@ -22,6 +22,8 @@ import org.gradle.api.Action;
 import org.gradle.api.specs.Spec;
 import org.gradle.internal.Actions;
 import org.gradle.internal.Cast;
+import org.gradle.internal.Factory;
+import org.gradle.internal.classloader.ClassLoaderUtils;
 import org.gradle.internal.exceptions.Contextual;
 import org.gradle.internal.exceptions.DefaultMultiCauseException;
 import org.gradle.internal.operations.BuildOperationExecutor;
@@ -52,6 +54,7 @@ import org.gradle.workers.WorkerExecutor;
 import org.gradle.workers.WorkParameters;
 import org.gradle.workers.WorkerSpec;
 
+import javax.annotation.Nullable;
 import javax.annotation.concurrent.NotThreadSafe;
 import java.io.File;
 import java.lang.reflect.ParameterizedType;
@@ -203,11 +206,11 @@ public class DefaultWorkerExecutor implements WorkerExecutor {
     private AsyncWorkCompletion submitWork(final ActionExecutionSpec spec, final IsolationMode isolationMode, final DaemonForkOptions daemonForkOptions) {
         final WorkerLease currentWorkerWorkerLease = getCurrentWorkerLease();
         final BuildOperationRef currentBuildOperation = buildOperationExecutor.getCurrentOperation();
+        WorkerFactory workerFactory = getWorkerFactory(isolationMode);
         WorkItemExecution execution = new WorkItemExecution(spec.getDisplayName(), currentWorkerWorkerLease, new Callable<DefaultWorkResult>() {
             @Override
             public DefaultWorkResult call() throws Exception {
                 try {
-                    WorkerFactory workerFactory = getWorkerFactory(isolationMode);
                     BuildOperationAwareWorker worker = workerFactory.getWorker(daemonForkOptions);
                     return worker.execute(spec, currentBuildOperation);
                 } catch (Throwable t) {
@@ -247,7 +250,7 @@ public class DefaultWorkerExecutor implements WorkerExecutor {
             case CLASSLOADER:
                 return isolatedClassloaderWorkerFactory;
             case NONE:
-                return noIsolationWorkerFactory;
+                return new ContextClassLoaderWorkerFactory(noIsolationWorkerFactory);
             case PROCESS:
                 return daemonWorkerFactory;
             default:
@@ -441,5 +444,36 @@ public class DefaultWorkerExecutor implements WorkerExecutor {
         public void await() throws WorkerExecutionException {
             workerExecutor.await(workItems);
          }
+    }
+
+    /**
+     * This is a delegating WorkerFactory that captures the context classloader at the moment it is created
+     * and then ensures that the worker is created with the same context classloader when
+     * getWorker() is called later.
+     */
+    private static class ContextClassLoaderWorkerFactory implements WorkerFactory {
+        private final ClassLoader contextClassLoader;
+        private final WorkerFactory delegate;
+
+        public ContextClassLoaderWorkerFactory(WorkerFactory delegate) {
+            this.delegate = delegate;
+            this.contextClassLoader = Thread.currentThread().getContextClassLoader();
+        }
+
+        @Override
+        public BuildOperationAwareWorker getWorker(DaemonForkOptions forkOptions) {
+            return ClassLoaderUtils.executeInClassloader(contextClassLoader, new Factory<BuildOperationAwareWorker>() {
+                @Nullable
+                @Override
+                public BuildOperationAwareWorker create() {
+                    return delegate.getWorker(forkOptions);
+                }
+            });
+        }
+
+        @Override
+        public IsolationMode getIsolationMode() {
+            return delegate.getIsolationMode();
+        }
     }
 }

--- a/subprojects/workers/src/testFixtures/groovy/org/gradle/workers/fixtures/WorkerExecutorFixture.groovy
+++ b/subprojects/workers/src/testFixtures/groovy/org/gradle/workers/fixtures/WorkerExecutorFixture.groovy
@@ -55,6 +55,9 @@ class WorkerExecutorFixture {
     }
 
     def prepareTaskTypeUsingWorker() {
+        withParameterClassInBuildSrc()
+        withFileHelperClassInBuildSrc()
+
         buildFile << """
             import org.gradle.workers.*
             $taskTypeUsingWorker
@@ -62,9 +65,6 @@ class WorkerExecutorFixture {
     }
 
     String getTaskTypeUsingWorker() {
-        withParameterClassInBuildSrc()
-        withFileHelperClassInBuildSrc()
-
         return """
             import javax.inject.Inject
             import org.gradle.other.Foo
@@ -241,7 +241,7 @@ class WorkerExecutorFixture {
             import java.io.FileWriter;
             
             public class FileHelper {
-                static void write(String id, File outputFile) {
+                public static void write(String id, File outputFile) {
                     PrintWriter out = null;
                     try {
                         outputFile.getParentFile().mkdirs();

--- a/subprojects/workers/workers.gradle.kts
+++ b/subprojects/workers/workers.gradle.kts
@@ -34,6 +34,7 @@ dependencies {
     integTestRuntimeOnly(project(":kotlinDsl"))
     integTestRuntimeOnly(project(":kotlinDslProviderPlugins"))
     integTestRuntimeOnly(project(":apiMetadata"))
+    integTestRuntimeOnly(project(":testKit"))
 
     integTestImplementation(project(":jvmServices"))
     integTestImplementation(project(":internalIntegTesting"))

